### PR TITLE
Math: Add Auditory functions library

### PR DESCRIPTION
**Pull request by singalsu**
The PR contains one commit for the library and another commit for cmocka tests.

### CODE REVIEW

**Comment by singalsu** [comment]
It's still work in progress, the Mel filterbank apply needs a test.

**Comment by lgirdwood** [comment]
@ShriramShastry pls can you review. Thanks !

**Comment by ShriramShastry** [comment]
> @ShriramShastry pls can you review. Thanks !

Sure, I 'II review the PR.
Thank you!!

**Comment by singalsu** [comment]
I just updated the PR with unit tests for Mel filterbank with various configurations:

```
sof/build_ut$ ctest --verbose -R auditory
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
Test project /home/singalsu/work/current/sof/sof/build_ut
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 52
    Start 52: auditory

52: Test command: /home/singalsu/work/current/sof/sof/build_ut/test/cmocka/src/math/auditory/auditory
52: Test timeout computed to be: 10000000
52: 1..10
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.96 (max  3.00)
52: ok 1 - test_mel_filterbank_16_test1
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.88 (max  3.00)
52: ok 2 - test_mel_filterbank_32_test1
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.47 (max  3.00)
52: ok 3 - test_mel_filterbank_16_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.51 (max  3.00)
52: ok 4 - test_mel_filterbank_32_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 5 - test_mel_filterbank_16_test3
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 6 - test_mel_filterbank_32_test3
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.22 (max  3.00)
52: ok 7 - test_mel_filterbank_16_test4
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.37 (max  3.00)
52: ok 8 - test_mel_filterbank_32_test4
52: Max absolute error =  1.00 (max  1.50), error RMS =  0.22 (max  0.50)
52: ok 9 - test_hz_to_mel
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.30 (max  1.50)
52: ok 10 - test_mel_to_hz
52: # ok - tests
1/1 Test #52: auditory .........................   Passed    0.07 sec

The following tests passed:
	auditory

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.07 sec

```

**Comment by singalsu** [comment]
The previous push with FFT patches included passed CI, so I removed them from the latest push. CI will pass after #6061 is merged.

File: src/include/sof/math/auditory.h
Comment (singalsu) [inline src/include/sof/math/auditory.h:39]: Fields scale and scale_inv can be converted to local variables. The correction for Mel triangles gain is done with scale_log2 now.

**Comment by singalsu** [comment]
New push fixed merge conflict in test/cmocka/src/math/CMakeLists.txt, no other changes.

File: test/cmocka/src/math/auditory/auditory.c
Comment (singalsu) [inline test/cmocka/src/math/auditory/auditory.c:39]: This needs to be #undef

File: test/cmocka/src/math/auditory/ref_auditory.m
Comment (singalsu) [inline test/cmocka/src/math/auditory/ref_auditory.m:9]: This line should not be commented to generate the test data

File: src/math/auditory/auditory.c
Comment (singalsu) [inline src/math/auditory/auditory.c:42]: Remove, this is no more used.

File: src/include/sof/math/auditory.h
Comment (singalsu) [inline src/include/sof/math/auditory.h:18]: No more used

File: src/math/auditory/auditory.c
Comment (singalsu) [inline src/math/auditory/auditory.c:62]: Some old code version in comment, remove

File: src/math/auditory/auditory.c
Comment (singalsu) [inline src/math/auditory/auditory.c:18]: Non used #define

File: src/math/auditory/auditory.c
Comment (singalsu) [inline src/math/auditory/auditory.c:31]: This is not used

**Review by singalsu — COMMENTED** [review]
Found some non-used definitions from old code version

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX this PR shouldn't fail on the DMIC test as it's not executed yet on that path. Can you check and confirm if OK to merge. Thanks !

File: src/math/auditory/auditory.c
Comment (lgirdwood) [inline src/math/auditory/auditory.c:45]: qualify for static inline i.e. will it get called a lot in hot code ?

**Review by lgirdwood — COMMENTED** [review]
LGTM, but lets confirm if some funcs need to be static inline.

**Comment by wszypelt** [comment]
@lgirdwood I ran the CI tests again

**Comment by lgirdwood** [comment]
@singalsu CI looks good, but some conflicts.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
license server issues..... re run the CI

File: test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h
Comment (singalsu) [inline test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h:6]: I'll do the same update as I did for matrix cmocka test vectors export. Would path "sof/test/cmocka/m" be a suitable for common Matlab language helper functions?

Comment (singalsu) [reply test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h]: Done now

File: src/math/auditory/auditory.c
Comment (singalsu) [reply src/math/auditory/auditory.c:45]: No, it's used only initialization. The Mel filterbank apply is the hot part. Did you find other inline candidates?

Comment (lgirdwood) [reply src/math/auditory/auditory.c:45]: no, but these only came to attention as they were small and simple.

**Comment by singalsu** [comment]
Reviewers, can you please review this.

File: test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m
Comment (ShriramShastry) [inline test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m:72]: can turn off them using `if isequal (variable_enable, flag)`

File: test/cmocka/src/math/auditory/mfcc_mel_to_hz.m
Comment (ShriramShastry) [inline test/cmocka/src/math/auditory/mfcc_mel_to_hz.m:13]: new line error ?

File: src/include/sof/math/auditory.h
Comment (ShriramShastry) [inline src/include/sof/math/auditory.h:51]: can rename `slaney`  flag?

**Review by ShriramShastry — COMMENTED** [review]
Code looks sound to me !!

**Comment by andrula-song** [comment]
LGTM~

File: src/include/sof/math/auditory.h
Comment (singalsu) [reply src/include/sof/math/auditory.h:51]: How, would slaney_normalize be better?

File: test/cmocka/src/math/auditory/mfcc_mel_to_hz.m
Comment (singalsu) [reply test/cmocka/src/math/auditory/mfcc_mel_to_hz.m:13]: yep it was missing

File: test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m
Comment (singalsu) [reply test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m]: Done, in beginning now plot_triangles and plot_matlab_triangles to enable the plots.

**Comment by singalsu** [comment]
> LGTM~

Don't hesitate to use the vote green "V" and red "X" buttons since you took the effort of reading the code!

**Comment by singalsu** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
